veomni/data/multimodal/preprocessor_registry.py

```diff
@@ -0,0 +1,153 @@
+
+        # Get the registered preprocessor for "my_custom_dataset"
+        preprocessor = get_preprocessor("my_custom_dataset")
+        result = preprocessor(conversations)
+"""
+
+from typing import Callable, Dict
+
+
+# Global registry for all dataset preprocessors
+_PREPROCESSOR_REGISTRY: Dict[str, Callable] = {}
+
+
+def register_preprocessor(name: str) -> Callable:
+    """
+    Decorator to register a dataset-specific preprocessor function.
+
+    Args:
+        name: Dataset source name (used as 'source_name' in your data config)
+
+    Returns:
+        Decorator function
+
+    Example:
+        @register_preprocessor("internal_vqa")
+        def internal_vqa_preprocess(conversations, **kwargs):
+            constructed_conversation = []
+            # ... your preprocessing logic
+            return constructed_conversation
+
+        # Can register multiple names for the same function:
+        @register_preprocessor("sharegpt4v_pretrain")
+        @register_preprocessor("sharegpt4v_captioner")
+        def sharegpt4v_pretrain_preprocess(conversations, **kwargs):
+            # ... preprocessing logic
+            return constructed_conversation
+    """
+
+    def decorator(func: Callable) -> Callable:
+        if name in _PREPROCESSOR_REGISTRY:
+            import warnings
+
+            warnings.warn(f"Preprocessor for '{name}' is already registered. Overwriting with new preprocessor.")
+        _PREPROCESSOR_REGISTRY[name] = func
+        return func
+
+    return decorator
+
+
+def get_all_preprocessors() -> Dict[str, Callable]:
+    """
+    Get all registered preprocessors.
+
+    This function ensures that all built-in preprocessors from preprocess.py
+    are loaded into the registry.
+
+    Returns:
+        Dictionary mapping dataset source names to preprocessor functions
+    """
+    # Import to trigger registration of all built-in preprocessors
+    # This import causes all @register_preprocessor decorators to execute
+    from . import preprocess  # noqa: F401
+
+    return _PREPROCESSOR_REGISTRY.copy()
+
+
+def get_preprocessor(name: str) -> Callable:
+    """
+    Get a specific preprocessor by dataset source name.
+
+    Args:
+        name: Dataset source name
+
+    Returns:
+        Preprocessor function
+
+    Raises:
+        ValueError: If preprocessor for the dataset name is not registered
+    """
+    # Ensure all preprocessors are loaded
+    get_all_preprocessors()
+
+    if name not in _PREPROCESSOR_REGISTRY:
+        raise ValueError(f"Unknown dataset name: {name}. No preprocessor registered for this source.")
+
+    return _PREPROCESSOR_REGISTRY[name]
+
+
+def list_preprocessors() -> list:
+    """
+    List all registered dataset source names.
+
+    Returns:
+        List of dataset source names (sorted)
+    """
+    get_all_preprocessors()
+    return sorted(_PREPROCESSOR_REGISTRY.keys())
+
+
+def is_preprocessor_registered(name: str) -> bool:
```

Comment:
I prefer make all the registry functions as a class. Check [chat_template_registry](https://github.com/ByteDance-Seed/VeOmni/blob/c57a165a36939431d5188a82d06b7569d307a704/veomni/data/chat_template.py#L252).
You can keep the decorator.